File: packages/server/src/launcher.ts
import { pathToFileURL } from "node:url";
import { readdir } from "node:fs/promises";
import path from "node:path";
import { withServerModuleDbScope } from "./db.ts";

const MODULE_EXTENSIONS = new Set([".ts", ".mts", ".js", ".mjs"]);

export interface DiscoverServerModulesOptions {
  modulesDir: string;
}

export interface DiscoveredServerModule {
  moduleName: string;
  filePath: string;
  methods: string[];
}

export interface ServerModuleInvocation {
  app: string;
  module: string;
  method: string;
  input?: unknown;
  requestId?: string;
  db?: Record<string, unknown>;
  meta?: {
    traceId?: string;
  };
}

export interface ServerModuleRegistry {
  listMethodsForModule(moduleName: string): string[];
  invoke(invocation: ServerModuleInvocation): Promise<unknown>;
}

export interface ServerModuleTransport {
  subscribe(subject: string, handler: (payload: unknown) => Promise<unknown>): Promise<void>;
}

export interface CreateServerModuleLauncherOptions extends DiscoverServerModulesOptions {
  app: string;
  transport: ServerModuleTransport;
}

export interface ServerModuleLauncher {
  start(): Promise<string[]>;
}

export interface StartServerModulesOptions extends CreateServerModuleLauncherOptions {}

export interface StartedServerModules {
  launcher: ServerModuleLauncher;
  subjects: string[];
}

type ModuleFunction = (input?: unknown) => unknown | Promise<unknown>;

interface LoadedServerModule {
  moduleName: string;
  methods: string[];
  handlers: Map<string, ModuleFunction>;
}

export function subjectForServerModule(app: string, moduleName: string, method: string): string {
  return `bladb.app.${app}.module.${moduleName}.${method}`;
}

export async function discoverServerModules(
  options: DiscoverServerModulesOptions,
): Promise<DiscoveredServerModule[]> {
  const entries = await readdir(options.modulesDir, { withFileTypes: true });
  const files = entries
    .filter((entry) => entry.isFile())
    .filter((entry) => !entry.name.startsWith("_"))
    .filter((entry) => MODULE_EXTENSIONS.has(path.extname(entry.name)))
    .sort((left, right) => left.name.localeCompare(right.name));

  const seen = new Map<string, string>();
  const discovered: DiscoveredServerModule[] = [];

  for (const file of files) {
    const filePath = path.join(options.modulesDir, file.name);
    const moduleName = path.basename(file.name, path.extname(file.name));
    const duplicate = seen.get(moduleName);
    if (duplicate) {
      throw new Error(
        `duplicate server module name \`${moduleName}\` for \`${duplicate}\` and \`${filePath}\``,
      );
    }

    seen.set(moduleName, filePath);
    const loaded = await loadServerModule(filePath, moduleName);
    discovered.push({
      moduleName,
      filePath,
      methods: loaded.methods,
    });
  }

  return discovered;
}

export async function createServerModuleRegistry(
  options: DiscoverServerModulesOptions,
): Promise<ServerModuleRegistry> {
  const entries = await readdir(options.modulesDir, { withFileTypes: true });
  const files = entries
    .filter((entry) => entry.isFile())
    .filter((entry) => !entry.name.startsWith("_"))
    .filter((entry) => MODULE_EXTENSIONS.has(path.extname(entry.name)))
    .sort((left, right) => left.name.localeCompare(right.name));

  const modules = new Map<string, LoadedServerModule>();

  for (const file of files) {
    const filePath = path.join(options.modulesDir, file.name);
    const moduleName = path.basename(file.name, path.extname(file.name));
    if (modules.has(moduleName)) {
      throw new Error(`duplicate server module name \`${moduleName}\``);
    }

    modules.set(moduleName, await loadServerModule(filePath, moduleName));
  }

  return {
    listMethodsForModule(moduleName: string) {
      return modules.get(moduleName)?.methods ?? [];
    },
    async invoke(invocation: ServerModuleInvocation) {
      const loaded = modules.get(invocation.module);
      if (!loaded) {
        throw new Error(`unknown server module \`${invocation.module}\``);
      }

      const handler = loaded.handlers.get(invocation.method);
      if (!handler) {
        throw new Error(
          `server module \`${invocation.module}\` does not export method \`${invocation.method}\``,
        );
      }

      return await withServerModuleDbScope(invocation.db, async () => {
        const previousPayload = Reflect.get(globalThis, "__bladbLauncherPayload");
        Reflect.set(globalThis, "__bladbLauncherPayload", invocation as unknown);
        try {
          return await handler(invocation.input);
        } finally {
          if (previousPayload === undefined) {
            Reflect.deleteProperty(globalThis, "__bladbLauncherPayload");
          } else {
            Reflect.set(globalThis, "__bladbLauncherPayload", previousPayload);
          }
        }
      });
    },
  };
}

export async function createServerModuleLauncher(
  options: CreateServerModuleLauncherOptions,
): Promise<ServerModuleLauncher> {
  const registry = await createServerModuleRegistry({ modulesDir: options.modulesDir });
  const discovered = await discoverServerModules({ modulesDir: options.modulesDir });

  return {
    async start() {
      const subjects: string[] = [];

      for (const moduleEntry of discovered) {
        for (const method of moduleEntry.methods) {
          const subject = subjectForServerModule(options.app, moduleEntry.moduleName, method);
          await options.transport.subscribe(subject, async (payload) => {
            const invocation = payload as ServerModuleInvocation;
            try {
              const data = await registry.invoke(invocation);
              return {
                ok: true,
                data,
                ...(invocation.requestId ? { requestId: invocation.requestId } : {}),
              };
            } catch (error) {
              const message = error instanceof Error ? error.message : "Unknown server module error";
              return {
                ok: false,
                error: {
                  code: "SERVER_MODULE_ERROR",
                  message,
                  module: invocation.module,
                  method: invocation.method,
                  ...(invocation.meta?.traceId ? { traceId: invocation.meta.traceId } : {}),
                },
                ...(invocation.requestId ? { requestId: invocation.requestId } : {}),
              };
            }
          });
          subjects.push(subject);
        }
      }

      return subjects.sort((left, right) => left.localeCompare(right));
    },
  };
}

export async function startServerModules(
  options: StartServerModulesOptions,
): Promise<StartedServerModules> {
  const launcher = await createServerModuleLauncher(options);
  const subjects = await launcher.start();
  return {
    launcher,
    subjects,
  };
}

async function loadServerModule(filePath: string, moduleName: string): Promise<LoadedServerModule> {
  const exportsObject = (await import(pathToFileURL(filePath).href)) as Record<string, unknown>;
  const handlers = new Map<string, ModuleFunction>();

  for (const [exportName, value] of Object.entries(exportsObject)) {
    if (exportName === "default") {
      continue;
    }

    if (typeof value !== "function") {
      continue;
    }

    handlers.set(exportName, value as ModuleFunction);
  }

  return {
    moduleName,
    methods: [...handlers.keys()].sort((left, right) => left.localeCompare(right)),
    handlers,
  };
}

File: crates/bladb-server/src/launcher.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::{collections::HashMap, sync::Arc};

use crate::transport::ServerModuleTransport;

pub type ServerModuleHandler =
    Arc<dyn Fn(ServerModuleInvocation) -> Result<Value, String> + Send + Sync + 'static>;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerModuleInvocation {
    pub app: String,
    pub module: String,
    pub method: String,
    #[serde(default)]
    pub input: Option<Value>,
    #[serde(default)]
    pub request_id: Option<String>,
    #[serde(default)]
    pub db: Option<Value>,
    #[serde(default)]
    pub meta: Option<Value>,
}

pub trait ServerModuleRegistry: Send + Sync {
    fn list_methods_for_module(&self, module_name: &str) -> Vec<String>;
    fn invoke(&self, invocation: ServerModuleInvocation) -> Result<Value, String>;
}

#[derive(Clone, Default)]
pub struct StaticServerModuleRegistry {
    modules: HashMap<String, HashMap<String, ServerModuleHandler>>,
}

impl StaticServerModuleRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register_handler(
        mut self,
        module_name: impl Into<String>,
        method: impl Into<String>,
        handler: ServerModuleHandler,
    ) -> Self {
        self.modules
            .entry(module_name.into())
            .or_default()
            .insert(method.into(), handler);
        self
    }
}

impl ServerModuleRegistry for StaticServerModuleRegistry {
    fn list_methods_for_module(&self, module_name: &str) -> Vec<String> {
        let Some(methods) = self.modules.get(module_name) else {
            return vec![];
        };

        let mut listed = methods.keys().cloned().collect::<Vec<_>>();
        listed.sort();
        listed
    }

    fn invoke(&self, invocation: ServerModuleInvocation) -> Result<Value, String> {
        let Some(module) = self.modules.get(&invocation.module) else {
            return Err(format!("unknown server module `{}`", invocation.module));
        };
        let Some(handler) = module.get(&invocation.method) else {
            return Err(format!(
                "server module `{}` does not export method `{}`",
                invocation.module, invocation.method
            ));
        };

        handler(invocation)
    }
}

pub struct CreateServerModuleLauncherOptions {
    pub app: String,
    pub transport: Arc<dyn ServerModuleTransport>,
    pub registry: Arc<dyn ServerModuleRegistry>,
    pub modules: Vec<String>,
}

pub trait ServerModuleLauncher: Send + Sync {
    fn start(&self) -> Result<Vec<String>, String>;
}

pub struct StaticServerModuleLauncher {
    app: String,
    transport: Arc<dyn ServerModuleTransport>,
    registry: Arc<dyn ServerModuleRegistry>,
    modules: Vec<String>,
}

impl ServerModuleLauncher for StaticServerModuleLauncher {
    fn start(&self) -> Result<Vec<String>, String> {
        let mut subjects = Vec::new();
        for module_name in &self.modules {
            for method in self.registry.list_methods_for_module(module_name) {
                let subject = subject_for_server_module(&self.app, module_name, &method);
                let registry = Arc::clone(&self.registry);
                self.transport.subscribe(
                    subject.clone(),
                    Arc::new(move |payload| {
                        let invocation: ServerModuleInvocation =
                            serde_json::from_value(payload).map_err(|error| error.to_string())?;
                        match registry.invoke(invocation.clone()) {
                            Ok(data) => Ok(json!({
                                "ok": true,
                                "data": data,
                                "requestId": invocation.request_id,
                            })),
                            Err(message) => Ok(json!({
                                "ok": false,
                                "error": {
                                    "code": "SERVER_MODULE_ERROR",
                                    "message": message,
                                    "module": invocation.module,
                                    "method": invocation.method,
                                },
                                "requestId": invocation.request_id,
                            })),
                        }
                    }),
                )?;
                subjects.push(subject);
            }
        }

        subjects.sort();
        Ok(subjects)
    }
}

pub struct StartedServerModules {
    pub subjects: Vec<String>,
}

pub fn create_server_module_launcher(
    options: CreateServerModuleLauncherOptions,
) -> Arc<dyn ServerModuleLauncher> {
    Arc::new(StaticServerModuleLauncher {
        app: options.app,
        transport: options.transport,
        registry: options.registry,
        modules: options.modules,
    })
}

pub fn start_server_modules(
    options: CreateServerModuleLauncherOptions,
) -> Result<StartedServerModules, String> {
    let launcher = create_server_module_launcher(options);
    let subjects = launcher.start()?;
    Ok(StartedServerModules { subjects })
}

pub fn subject_for_server_module(app: &str, module_name: &str, method: &str) -> String {
    format!("bladb.app.{app}.module.{module_name}.{method}")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::create_in_memory_server_module_transport;

    #[test]
    fn registers_subjects_and_invokes_handlers() {
        let transport = Arc::new(create_in_memory_server_module_transport());
        let registry = Arc::new(StaticServerModuleRegistry::new().register_handler(
            "user",
            "login",
            Arc::new(|invocation| {
                let email = invocation
                    .input
                    .and_then(|value| value.get("email").cloned())
                    .and_then(|value| value.as_str().map(str::to_string))
                    .ok_or_else(|| "missing email".to_string())?;
                Ok(json!({ "email": email, "token": "session-demo" }))
            }),
        ));

        let started = start_server_modules(CreateServerModuleLauncherOptions {
            app: "demo".into(),
            transport: transport.clone(),
            registry,
            modules: vec!["user".into()],
        })
        .expect("start rust server modules");

        assert_eq!(
            started.subjects,
            vec!["bladb.app.demo.module.user.login".to_string()]
        );

        let response = transport
            .request(
                "bladb.app.demo.module.user.login",
                json!({
                    "app": "demo",
                    "module": "user",
                    "method": "login",
                    "input": {
                        "email": "member@example.com"
                    }
                }),
            )
            .expect("invoke in-memory transport");

        assert_eq!(response["ok"], true);
        assert_eq!(response["data"]["email"], "member@example.com");
    }
}
